### PR TITLE
Update GDScript 2.0 engine to use weak references for scripts and packed scenes to enable cyclic references

### LINKED ISSUE #21461: Some usages of `class_name` can produce cyclic errors unexpectedly
___

***Bugsquad edit:** This issue has been confirmed several times already. No need to confirm it further.*

___

Windows 10 64-bit - Godot 3.1 alpha

## Edit - Refined issue to focus on cyclic errors.

Certain uses of `class_name` can produce cyclic errors in situations where there are no cyclic references, and in fact sometimes no other references outside of the single one given.

I've noticed this most frequently in tool script when using `is` to make a class comparison, but it seems to happen at times in normal script situations.

Recreating it in a sample project has eluded me.

godot master_2018-08-27_00-47-02

### LINKED ISSUE #32165: Cyclic Errors After Autoload of Imported Asset
**Godot version:** 3.1 (confirmed)

**OS/device including version:** Windows 10 & Linux

**Issue description:**

I recently had an Asset approved on the Asset Lib. I was disappointed that trying my new Asset using the Import system produced errors. There is probably another request I want to open regarding having a TEST site for Assets, but for now I want to report this.

After importing my Asset (Godot-Stuff Logging 3.1) and setting up and Autoload global, I receive a number of errors as seen below.

The Workaround is to close your project and reopen it.

```
 Resource: 'res://gs_logger/logger.gd' is already being loaded. Cyclic reference?
 res://gs_logger/layouts/pattern_layout.gd:13 - Parse Error: Identifier 'Logger' is not declared in the current scope.
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERROR
 res://gs_logger/appender.gd:53 - Parse Error: Class 'PatternLayout' could not be fully loaded (script error or cyclic dependency).
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERROR
 res://gs_logger/logger.gd:268 - Parse Error: Class 'Appender' could not be fully loaded (script error or cyclic dependency).
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERROR
 Script does not inherit a Node: res://gs_logger/logger.gd
 editor/editor_autoload_settings.cpp:448 - Condition ' !info->node ' is true. Continuing..:
Add AutoLoad
```

**Steps to reproduce:**

- create a new project
- use the assetlib and download and import Godot-Stuff 3.1 Logger
- create a new Autoload for the script gs_logger/logger.gd and call it "Logger"
- you will receive the errors above
- create a new scene and attach a gdscript to it
- try to access the Logger global object you will receive the following errors

```
 Resource: 'res://gs_logger/logger.gd' is already being loaded. Cyclic reference?
 res://gs_logger/layouts/pattern_layout.gd:13 - Parse Error: Identifier 'Logger' is not declared in the current scope.
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERROR
 res://gs_logger/appender.gd:53 - Parse Error: Class 'PatternLayout' could not be fully loaded (script error or cyclic dependency).
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERROR
 res://gs_logger/logger.gd:268 - Parse Error: Class 'Appender' could not be fully loaded (script error or cyclic dependency).
 modules/gdscript/gdscript.cpp:580 - Method/Function Failed, returning: ERR_PARSE_ERRO
```
- close the project and re-open it
- the errors disappear

**Minimal reproduction project:**

### LINKED ISSUE #41027: Cyclic reference error in GDScript 2.0
**Godot version:**
Godot Eninge v4.0.dev.custom_build.ec678c5e9

**OS/device including version:**
Windows 10

**Issue description:**
Cyclic reference error:
image

**Steps to reproduce:**
1) Create class1 and class2
2) Initialize class1 in class2 and class2 in class1
3) Add class1 to scene

**Minimal reproduction project:**
[godot_cyclic_reference.zip](https://github.com/godotengine/godot/files/5023741/godot_cyclic_reference.zip)

### LINKED ISSUE #58181: GDScript 2.0: Cannot preload cyclic const script references
### Godot version

b3507ab0cdf25e470b1e4d977ed0bea32a305c1b

### System information

Windows 10, Vulkan, Intel® Core™ i7-8700K, Nvidia GTX 1080 (471.11)

### Issue description

Scripts can preload other scripts with the 'preload' keyword, however, they will fail with the error ```Constant value uses script from "%s" which is loaded but not compiled``` if trying to preload a script which itself has preload keywords for the script which is loading, meaning it cannot resolve this cyclic dependency.

There are TODO messages in the ```GDScriptAnalyzer::reduce_preload``` which seem to hint at a strategy to support cyclic preloads, but it is unclear how this should be accomplished in the current GDScript framework. Requesting information from @vnen on the best way to potentially resolve this.

### Steps to reproduce

Simply create two scripts which use the ```preload``` function to preload each other.

### Minimal reproduction project

[cyclic_preload.zip](https://github.com/godotengine/godot/files/8079475/cyclic_preload.zip)

### LINKED ISSUE #58551: Can't preload a scene in AutoLoad script when the scene's script references the AutoLoad
### Godot version

4.0.alpha3

### System information

Windows 10

### Issue description

This appears to be a regression in 4.0, as this works in 3.4.3. Running the scene causes a ParserError crash, when I expect it to run fine.

### Steps to reproduce

If you have a simple AutoLoad script like so:
image

Which preloads this simple scene with just one script:
image

And that script references the AutoLoad script:
image

Then running the scene causes a crash:
`Parser Error: Could not preload resource file "res://Scene.tscn"`

### Minimal reproduction project

Working project in 3.4.3:
[AutoloadBugGodot3.4.3.zip](https://github.com/godotengine/godot/files/8146324/AutoloadBugGodot3.4.3.zip)

Identical project but in 4.0.alpha3:
[AutoloadBugGodot4.zip](https://github.com/godotengine/godot/files/8146325/AutoloadBugGodot4.zip)

### LINKED ISSUE #58871: Cyclic Autoload Singleton Plugin Bug.
### Godot version

v3.4.3.stable.official [242c05d12]

### System information

Windows 10

### Issue description

Cannot enable plugin because autoload is not declared. Autoload is not declared because plugin is disable.

Reloading project does not help.
You have to remove all the singleton reference to temporary fix the bug.

### Steps to reproduce

1. Make a plugin.
2. Create a singleton script and add to plugin.
3. Create any script that reference to that singleton and add to plugin.
4. Disable the plugin and reload project.
5. Try to enable the plugin and now you can't until you remove that reference.

### Minimal reproduction project

[Cyclic Autoload Singleton Plugin Bug.zip](https://github.com/godotengine/godot/files/8199572/Cyclic.Autoload.Singleton.Plugin.Bug.zip)

### LINKED ISSUE #61043: New cyclic reference error with autoload and preloaded script
### Godot version

6d5d9dd

### System information

Windows 10 x64

### Issue description

I have a `class_name` script `Game` that preloads another script that uses an autoload that uses a variable of the type `Game`. There is obvious cyclic reference here, but it runs fine in 3.5. In 4.0 the script is parsed properly in the editor, but at runtime it stops on `Parser Error: Could not preload resource file` at the `preload` line.

### Steps to reproduce

The steps are confusing, just check the project.

### Minimal reproduction project

[AutoloadBug.zip](https://github.com/godotengine/godot/files/8694166/AutoloadBug.zip)
You can run it both in 3.5 and 4.0 and compare results.

### PR DESCRIPTION
**Pull request by adamscott**
_Superseded by #67714_

## TL;DR
This PR makes possible to use cyclic references in GDScript 2.0

## Summary
This PR makes the GDScript 2.0 engine compatible with weak references of Scripts and GDScripts. Instead of passing the scripts around, which makes the counter up, this PR stores the scripts inside their own weak references. So, the scripts can be used as much as needed without making the counter up. This makes cyclic references possible.

## What it tries to solve
The current GDScript engine uses strong references to store the scripts and scenes that it loads. This usually work well in the standard situations, as those scripts and scenes are `RefCounted` objects.

### Standard case
Let's say `Main` loads `A` that loads `B`.

Here, `Main` reference count is 1 (referenced by the engine), `A` is 1 (referenced by `Main`) and `B` is 1 (referenced by `A`).

When the engine unloads, `Main` reference count will be decremented by 1 and will be deleted, because it's reference count will hit 0. `A` and `B` will also be deleted because `A` lost his reference from `Main` and `B` will lose it's reference from `A` later.

### Current cyclic reference issue
Currently, the engine makes it so that using cyclic references is impossible. Why, you say?

Let's bring back our example, but with a cyclic reference. Let's say `Main` loads `A` that loads `B` that loads back `A`.

Here, `Main` reference count is 1 (referenced by the engine), `A` is 2 (referenced by `Main` and `B`) and `B` is 1 (referenced by `A`).

Do you see the issue? When `Main` will be deleted, there will be no way to delete `A` and `B`. `A` will lose it's reference to `Main` and will decrement, but it will not be deleted as `A` reference count is now 1. `B` is not deleted because it didn't lose it's reference to `A`.

So, the answer is: the engine makes cyclic references impossible due to memory leaks concerns. That explains errors thrown by the parser or the analyzer like `Constant value uses script from "res://b.gd" which is loaded but not compiled`.

## Proposed solution
Let's bring back our example.

`Main` loads `A` that loads `B` that loads back `A`.

What if loading increments a wrapper instead of the script reference? And what if we use a dependency chart to detect when a script needs to be deleted?

### Weak references
Instead of using strong references for scripts and scenes, this PR uses weak references. Weak references are referenced values that holds another value. When accessing that later value, it doesn't increase the reference counter, making it so that it does not matter how cyclic references are. 

Instead of having a strong reference to `A` in `B` (which would increase `A` to 2), let's use altogether weak references for scripts and scenes. `B` will still be able to use `A`, but `A` reference count will stay at 1.

The engine had to be changed to support values that may be a weak reference or not. Hence the rather heavy PR.

### Dependency chart
If we don't use references as the way to dispose of the loaded objects, we need another way to delete them. Here comes the dependency chart.

Take the dependency chart of our example
- `Main`
  - loads `A`
- `A`
  - loads `B`
- `B`
  - loads `A`

With this, we know that `A` and `B` cannot be unloaded. Because `A` and `B` are found in other dependencies. But when `Main` will be disposed of, it's only a matter of disposing of every linked dependency.

## List of changes:
-  Add weak references to support preload cyclic references
-  Add new algorithm for removing scripts based on dependencies
-  Expose `GDScriptCache::remove_script()`
-  Use `.notest.gd` system to bypass tests
-  Add `ScriptRef`
-  Add weakref for `GDScriptParser`
-  Add global classes support (replaces #65664)
-  Add packed scenes support

## Known issues
- heap-use-after-free or leaks issues

## Code example
So, this code works with this PR, which returns this error on `master`: `Constant value uses script from "res://b.gd" which is loaded but not compiled.`:
```gdscript
# main.gd
const A = preload("res://a.gd")

func _ready() -> void:
	A.test()  # The console will print: "A.HELLO_WORLD: hello, world!"
```
```gdscript
# a.gd
extends Node

const B = preload("res://b.gd")

const HELLO_WORLD: = "hello, world!"

static func test() -> void:
	B.test()

	# Test cyclic `is` keyword
	var b_instance: = B.new()
	if b_instance is B:
		prints("b_instance is B")  # The console will print: "b_instance is B"
	b_instance.queue_free()
```
```gdscript
# b.gd
extends Node

const A = preload("res://a.gd")

static func test() -> void:
	prints("A.HELLO_WORLD:", A.HELLO_WORLD)

	# Test cyclic `is` keyword
	var a_instance: = A.new()
	if a_instance is A:
		prints("a_instance is A")  # The console will print: "a_instance is A"
	a_instance.queue_free()
```

## Minimal reproduction project:
[cyclic3.zip](https://github.com/godotengine/godot/files/9543713/cyclic3.zip)

## Fixes
Fixes #21461 - Some usages of `class_name` can produce cyclic errors unexpectedly
Fixes #32165 - Cyclic Errors After Autoload of Imported Asset
Fixes #41027 - Cyclic reference error in GDScript 2.0
Fixes #58551 - Can't preload a scene in AutoLoad script when the scene's script references the AutoLoad
Fixes #58181 - GDScript 2.0: Cannot preload cyclic const script references
Fixes #58871 - Cyclic Autoload Singleton Plugin Bug
Fixes #61043 - New cyclic reference error with autoload and preloaded script

## Notes
This PR was the follow-up to #65672.
This PR replaces #65664 as it implements global classes too.

### CODE REVIEW

**Comment by KoBeWi** [comment]
I tested it with my project and got this crash:
```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.0.beta.custom_build (f74491fdee9bc2d68668137fbacd8f3a7e7e8df7)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] ObjectID::is_null (C:\godot_source\core\object\object_id.h:47)
[1] ObjectID::is_null (C:\godot_source\core\object\object_id.h:47)
[2] WeakRef::get_ref (C:\godot_source\core\object\ref_counted.cpp:103)
[3] ScriptRef::get_ref (C:\godot_source\core\object\script_language.h:258)
[4] GDScriptRef::get_ref (C:\godot_source\modules\gdscript\gdscript.h:540)
[5] ResourceFormatLoaderGDScript::load (C:\godot_source\modules\gdscript\gdscript.cpp:2402)
[6] ResourceLoader::_load (C:\godot_source\core\io\resource_loader.cpp:213)
[7] ResourceLoader::_thread_load_function (C:\godot_source\core\io\resource_loader.cpp:240)
[8] ResourceLoader::load (C:\godot_source\core\io\resource_loader.cpp:570)
[9] EditorAutoloadSettings::_create_autoload (C:\godot_source\editor\editor_autoload_settings.cpp:403)
[10] EditorAutoloadSettings::EditorAutoloadSettings (C:\godot_source\editor\editor_autoload_settings.cpp:860)
[11] ProjectSettingsEditor::ProjectSettingsEditor (C:\godot_source\editor\project_settings_editor.cpp:678)
[12] EditorNode::EditorNode (C:\godot_source\editor\editor_node.cpp:6664)
[13] Main::start (C:\godot_source\main\main.cpp:2754)
[14] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:176)
[15] _main (C:\godot_source\platform\windows\godot_windows.cpp:201)
[16] main (C:\godot_source\platform\windows\godot_windows.cpp:215)
[17] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:229)
[18] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[19] BaseThreadInitThunk
-- END OF BACKTRACE --
```

**Comment by adamscott** [comment]
@KoBeWi Thanks! This should be fixed by now.

**Comment by KoBeWi** [comment]
Another crash, this time I didn't get any stack trace, had to use debugger to find it:
```
CowData<char32_t>::size() Line 133 (c:\godot_source\core\templates\cowdata.h:133)
CowData<char32_t>::get(int p_index) Line 157 (c:\godot_source\core\templates\cowdata.h:157)
String::operator[](int p_index) Line 224 (c:\godot_source\core\string\ustring.h:224)
String::get_data() Line 970 (c:\godot_source\core\string\ustring.cpp:970)
String::hash() Line 2688 (c:\godot_source\core\string\ustring.cpp:2688)
HashMapHasherDefault::hash(const String & p_string) Line 306 (c:\godot_source\core\templates\hashfuncs.h:306)
HashMap<String,Ref<GDScript>,HashMapHasherDefault,HashMapComparatorDefault<String>,DefaultTypedAllocator<HashMapElement<String,Ref<GDScript>>>>::_hash(const String & p_key) Line 85 (c:\godot_source\core\templates\hash_map.h:85)
HashMap<String,Ref<GDScript>,HashMapHasherDefault,HashMapComparatorDefault<String>,DefaultTypedAllocator<HashMapElement<String,Ref<GDScript>>>>::_lookup_pos(const String & p_key, unsigned int & r_pos) Line 106 (c:\godot_source\core\templates\hash_map.h:106)
HashMap<String,Ref<GDScript>,HashMapHasherDefault,HashMapComparatorDefault<String>,DefaultTypedAllocator<HashMapElement<String,Ref<GDScript>>>>::erase(const String & p_key) Line 315 (c:\godot_source\core\templates\hash_map.h:315)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 136 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:136)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
GDScriptCache::remove_dependencies(const String & p_source, const String & p_path, const bool & repeat) Line 141 (c:\godot_source\modules\gdscript\gdscript_cache.cpp:141)
```

**Comment by h0lley** [comment]
Loaded a medium sized project with a2a9817.
no crashes, but got a bunch of
`Attempt to open script 'res://my_scene.tscn::1' resulted in error 'File not found'.`
it seems be built-in scripts (scripts as sub-resources) only.

then when starting test play, I got this error:
`Invalid call. Nonexistent function 'new' in base 'GDScript'.`

**Comment by adamscott** [comment]
@h0lley This should be fixed with [6c708e9](https://github.com/godotengine/godot/pull/65752/commits/6c708e92fc4ebaca24ef78ea8ee5accb0e73d0fc), at least the bunch of complaints about internal scripts.

**Comment by bfeber** [comment]
This doesn't actually fix the bug from the first MRP here: https://github.com/godotengine/godot/issues/61043#issuecomment-1257244152

**Comment by h0lley** [comment]
tested again with aef6be9.
`Attempt to open script resulted in error 'File not found'.` errors are gone!

the `Invalid call. Nonexistent function 'new' in base 'GDScript'.` error still persists, however this may be a separate issue reported in #51538. [this workaround](https://github.com/godotengine/godot/issues/51538#issuecomment-1063494798) no longer works on your branch, however the [workaround mentioned by KoBeWi](https://github.com/godotengine/godot/issues/51538#issuecomment-1140529421) still does, so I guess it's fine to ignore for the time being.

I then proceeded to test code that would previously cause a cyclic reference error, which at first looked like it was resolved now, however on runtime, I then got a `Right operand of 'is' is not a class.` error.

here are minimal reproduction steps:
1. create two class scripts:
```swift
extends Node
class_name TestA

var test: Node

func _ready():
	if test is TestB:
		pass
```
```swift
extends Node
class_name TestB

var test: Node

func _ready():
	if test is TestA:
		pass
```
2. create a scene and add both of them as nodes
3. F6 launch into the scene

**Comment by adamscott** [comment]
> This doesn't actually fix the bug from the first MRP here: [#61043 (comment)](https://github.com/godotengine/godot/issues/61043#issuecomment-1257244152)

Yes. And I am pretty much thinking about creating a shallow packed scene loading system. But it requires thinking about it a bit.

**Comment by adamscott** [comment]
The problem I have right now, is scene preloading. I think my PR is stable enough for scripts, but cyclic scene preloading brings it's own issues.

The problem with the cyclic preload of scenes is that the scene is preloading itself. To preload scenes, the engine uses `ResourceLoader::load`, which loads and parse the scene, but the code itself parses the same preload which calls again `ResourceLoader::load` for the same scene (remember, the first load isn't done, it's currently running waiting to finish the parse).

Like I mentionned, I think that a "shallow" scene is needed, just like shallow scripts exists. Without shallow scripts, it would be impossible for cyclic references. We need a way to create a shell of a scene which is accessed on runtime.

So, the issue is not what to do, but how to do it.

**Comment by h0lley** [comment]
not a maintainer here but beware of feature creep. if your PR tries to implement too many things at once, you may get told to split it up.

as an user, for starters I'd be exhilarated to finally be able to reference my classes static properties anywhere, particularly to access enums they declare or check with the `is` keyword.

I'm not sure how common of a problem cyclic scene (or any Resource for that matter) reference really is.

**Comment by KoBeWi** [comment]
Yeah, if a PR changes too many things at once it makes it difficult to review. It's better to split it up *if possible*.

**Comment by adamscott** [comment]
> This doesn't actually fix the bug from the first MRP here: [#61043 (comment)](https://github.com/godotengine/godot/issues/61043#issuecomment-1257244152)

This should be fixed by now.

**Comment by h0lley** [comment]
with 4d3b0f2 the issues with `is` I reported earlier seem to be resolved!

I've encountered a new issue though.
1. create two scenes
2. in one of them, attach a script:
```swift
extends Node2D

const dict = {"scene": preload("res://other_scene.tscn")}

func _ready():
	var key = "scene"
	dict[key].instantiate()
```
3. F6 launch into it:
```
Invalid call. Nonexistent function 'instantiate' in base 'WeakRef'.
```
the issue does not occur when the key to the Dictionary isn't variable, i.e. when using `dict["scene"]` or `dict.scene`, nor does it occur when the Dictionary isn't const.

**Comment by adamscott** [comment]
> with [4d3b0f2](https://github.com/godotengine/godot/commit/4d3b0f25e56480e771147480f753006922508aa5) the issues with `is` I reported earlier seem to be resolved!

That's great, because I didn't had the time to check this out!

**Comment by adamscott** [comment]
@h0lley I fixed the dict access, but I created a leak. Or maybe, the leak was bound to happen anyway. Therefore, I'm again in the process to fix those pesky leaks.

**Comment by h0lley** [comment]
compiled 065aaa5 and can confirm [this issue](https://github.com/godotengine/godot/pull/65752#issuecomment-1261581447) now fixed! however now when trying to test this [this one](https://github.com/godotengine/godot/pull/65752#issuecomment-1257283580) again I am getting a crash:
```
================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.0.beta.custom_build (065aaa50c65c6e46619cafbede63b87eebe9ac04)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f655d3b7520] (??:0)
[2] GDScriptAnalyzer::is_type_compatible(GDScriptParser::DataType const&, GDScriptParser::DataType const&, bool, GDScriptParser::Node const*) (??:0)
[3] GDScriptAnalyzer::reduce_binary_op(GDScriptParser::BinaryOpNode*) (??:0)
[4] GDScriptAnalyzer::reduce_expression(GDScriptParser::ExpressionNode*, bool) (??:0)
[5] GDScriptAnalyzer::resolve_if(GDScriptParser::IfNode*) (??:0)
[6] GDScriptAnalyzer::resolve_node(GDScriptParser::Node*) (??:0)
[7] GDScriptAnalyzer::resolve_suite(GDScriptParser::SuiteNode*) (??:0)
[8] GDScriptAnalyzer::resolve_function_body(GDScriptParser::FunctionNode*) (??:0)
[9] GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode*) (??:0)
[10] GDScriptAnalyzer::resolve_program() (??:0)
[11] GDScriptAnalyzer::analyze() (??:0)
[12] GDScript::reload(bool) (??:0)
[13] GDScriptCache::get_full_script(String const&, Error&, String const&, bool) (??:0)
[14] ResourceFormatLoaderGDScript::load(String const&, String const&, Error*, bool, float*, ResourceFormatLoader::CacheMode) (??:0)
[15] ResourceLoader::_load(String const&, String const&, String const&, ResourceFormatLoader::CacheMode, Error*, bool, float*) (??:0)
[16] ResourceLoader::_thread_load_function(void*) (??:0)
[17] ResourceLoader::load(String const&, String const&, ResourceFormatLoader::CacheMode, Error*) (??:0)
[18] EditorFileSystem::_scan_new_dir(EditorFileSystemDirectory*, Ref<DirAccess>&, EditorFileSystem::ScanProgress const&) (??:0)
[19] EditorFileSystem::_scan_filesystem() (??:0)
[20] EditorFileSystem::_thread_func(void*) (??:0)
[21] Thread::callback(Thread*, Thread::Settings const&, void (*)(void*), void*) (??:0)
[22] void std::__invoke_impl<void, void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(std::__invoke_other, void (*&&)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*&&, Thread::Settings&&, void (*&&)(void*), void*&&) (??:0)
[23] std::__invoke_result<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>::type std::__invoke<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*&&, Thread::Settings&&, void (*&&)(void*), void*&&) (??:0)
[24] void std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (??:0)
[25] std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::operator()() (??:0)
[26] std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > >::_M_run() (??:0)
[27] /media/holly/Workspace/Projects/Godot/godot/bin/godot.linuxbsd.tools.x86_64.llvm.dev.tmpcommit(+0x676e6b4) [0x55785ef136b4] (??:0)
[28] /lib/x86_64-linux-gnu/libc.so.6(+0x94b43) [0x7f655d409b43] (??:0)
[29] /lib/x86_64-linux-gnu/libc.so.6(+0x126a00) [0x7f655d49ba00] (??:0)
-- END OF BACKTRACE --
================================================================
```
sorry for no file names and line numbers in the backtrace, I don't know how to get those.

**Comment by bitbrain** [comment]
Trying to open my game with the latest changes crashes Godot - these errors do not happen with current `master`:
```
SCRIPT ERROR: Parse Error: Could not parse global class "MiningItem" from "res://items/equipment/MiningItem.gd".
          at: GDScript::reload (res://savegame/SaveGame.gd:57)
ERROR: Script does not inherit from Node: res://savegame/SaveGame.gd.
   at: (editor\editor_autoload_settings.cpp:413)
SCRIPT ERROR: Parse Error: Could not parse global class "WorldCharacter" from "res://worldobjects/character/WorldCharacter.gd".
          at: GDScript::reload (res://ai/behaviors/actions/CarryItemAction.gd:26)
SCRIPT ERROR: Parse Error: Could not parse global class "BeehaveRoot" from "res://addons/beehave/nodes/beehave_root.gd".
          at: GDScript::reload (res://ui/debug/DebugInfo.gd:9)
SCRIPT ERROR: Parse Error: Could not parse global class "BeehaveRoot" from "res://addons/beehave/nodes/beehave_root.gd".
          at: GDScript::reload (res://ui/debug/DebugInfo.gd:72)
SCRIPT ERROR: Parse Error: Could not parse global class "BeehaveRoot" from "res://addons/beehave/nodes/beehave_root.gd".
          at: GDScript::reload (res://ui/debug/DebugInfo.gd:82)
SCRIPT ERROR: Parse Error: Could not parse global class "BeehaveRoot" from "res://addons/beehave/nodes/beehave_root.gd".
          at: GDScript::reload (res://ui/debug/DebugInfo.gd:22)
SCRIPT ERROR: Parse Error: Could not parse global class "InteractionController" from "res://interaction/InteractionController.gd".
          at: GDScript::reload (res://ui/debug/DebugInfo.gd:43)
SCRIPT ERROR: Parse Error: Could not parse global class "Inventory" from "res://items/Inventory.gd".
          at: GDScript::reload (res://worldobjects/character/WorldCharacter.gd:122)
SCRIPT ERROR: Parse Error: Could not parse global class "Equipment" from "res://items/Equipment.gd".     
          at: GDScript::reload (res://worldobjects/character/WorldCharacter.gd:125)
SCRIPT ERROR: Parse Error: Could not parse global class "Character" from "res://characters/Character.gd".
SCRIPT ERROR: Parse Error: Could not parse global class "ItemSpawner" from "res://items/ItemSpawner.gd".
          at: GDScript::reload (res://ai/behaviors/actions/DropItemAction.gd:7)
SCRIPT ERROR: Parse Error: Could not parse global class "Inventory" from "res://items/Inventory.gd".
          at: GDScript::reload (res://ai/behaviors/actions/DropItemAction.gd:44)
SCRIPT ERROR: Parse Error: Could not parse global class "Inventory" from "res://items/Inventory.gd".
          at: GDScript::reload (res://ai/behaviors/actions/DropItemAction.gd:14)
SCRIPT ERROR: Parse Error: Could not parse global class "Inventory" from "res://items/Inventory.gd".
          at: GDScript::reload (res://ai/behaviors/actions/DropItemAction.gd:15)
ERROR: Parser bug (please report): Trying to check compatibility of unset target type
   at: (modules\gdscript\gdscript_analyzer.cpp:4115)
SCRIPT ERROR: Parse Error: Could not parse global class "MapData" from "res://maps/MapData.gd".
          at: GDScript::reload (res://maps/CaveGenerator.gd:31)
SCRIPT ERROR: Parse Error: Could not parse global class "Chest" from "res://worldobjects/chest/Chest.gd".
          at: GDScript::reload (res://scenes/levels/ProceduralCaveScene.gd:65)
```
- `MiningItem` extends a custom resource `Item`that extends `Resource`
- things like MapData and Chest are script classes defined via class_name

**Comment by bitbrain** [comment]
Here are similar errors on this small project:

[dependency-errors.zip](https://github.com/godotengine/godot/files/9689994/dependency-errors.zip)

This does not crash Godot though, it looks like that these changes in combination with other things (like singleton nodes or addons) might let it crash. Still, these errors do not appear on current `master`

**Comment by adamscott** [comment]
Most bugs should be resolved by now. @KoBeWi graciously gave me access to his (huge) personal project and it loaded without crashing.

Don't hesitate to post bugs or crashes you find. Thanks again!

**Comment by bitbrain** [comment]
I opened this in my Godot 4 ported RPG (around 2 years of development on it already) and it gave me no errors either! (apart from regression errors that are already covered by other issues)

**Comment by h0lley** [comment]
for my project any of the errors / crashes I posted earlier in this thread have been resolved as well!
however there are still a couple of errors being printed, although cannot link them to anything malfunctioning.

I get this one when loading the project in the editor or when starting test play:
```
E 0:00:01:0015   _ready: Attempt to open script '' resulted in error 'File not found'.
  <C++ Error>    Condition "err" is true. Returning: err
  <C++ Source>   modules/gdscript/gdscript.cpp:1065 @ load_source_code()
  <Stack Trace>  boot.tscn::1:40 @ _ready()
```
the file in question appears to be a built-in script attached to the main scene.
line 40 in said script contains `get_tree().change_scene_to_file("...")`.
unfortunately I haven't been able to reconstruct it in a minimal project so far.

then, when dragging GD script files that register classes between folders in the FileSystem panel, I get these errros:
```
  Attempt to open script 'res://cyclic/class_a.gd' resulted in error 'File not found'.
  res://cyclic/class_a.gd:2 - Parse Error: Class "TestA" hides a global script class.
```
this can be easily reconstructed.
for instance in this project: [tmp.zip](https://github.com/godotengine/godot/files/9709402/tmp.zip)
1. drag the `class_a.gd` from `cyclic` into the `res://`
2. notice error: `res://class_a.gd:2 - Parse Error: Class "TestA" hides a global script class.`
3. drag it back into `cyclic`
4. notice error: `Attempt to open script 'res://class_a.gd' resulted in error 'File not found'.`

**Comment by bitbrain** [comment]
Hopefully this makes it into Beta 3 *keeps fingers crossed*

File: modules/gdscript/gdscript_cache.cpp [hunk 1]
```diff
@@ -394,7 +394,19 @@ GDScriptCache::~GDScriptCache() {
 	packed_scene_cache.clear();
 
 	parser_map.clear();
+
+	for (KeyValue<String, Ref<GDScript>> &E : shallow_gdscript_cache) {
```
[the remark this answers is not in the document]
Comment (adamscott) [reply modules/gdscript/gdscript_cache.cpp:398]: This code was removed, as this was quite a ugly hack. Now, as references are not leaking, clearing `shallow_gdscript_cache` and `full_gdscript_cache` suffice.

[the remark this answers is not in the document]
Comment (adamscott) [reply modules/gdscript/gdscript_cache.cpp]: This was fixed!

**Comment by adamscott** [comment]
Superseded by #67714
